### ISSUE TITLE
TruffleBoundary#throwsControlFlowException() is badly documented

### ISSUE THREAD

**Issue by eregon** [oracle/graal#246]
Quoting the [source](https://github.com/graalvm/graal/blob/992fea2c78035d9248be707f7d68155134441d56/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/CompilerDirectives.java#L241-L245):
```java
    public @interface TruffleBoundary {
        /**
         * Determines whether this method throws a {@link ControlFlowException}.
         */
        boolean throwsControlFlowException() default false;
```

But this is wrong, as this seems needed for any unchecked exception thrown by the method, not just ControlFlowException. So it should be renamed.

Also, it doesn't mention anything about what it does.
In fact, if you miss this flag, the method will deoptimize on __*every*__ exception thrown by the method, which has a huge cost on performance.

This is actually very unfortunate because:
* We have to specify this manually. This is extremely easy to forget and very hard to say if a method can throw an exception (basically every single method could potentially throw an exception like ClassCastException or NullPointerException) and one need to inspect all possible (transitive!) sub-calls from that method (which is unreasonable, usually things beyond a boundary involve many calls). 
* Adding the flag complexifies the graph in a non-trivial way by adding an exception path which has a non-negligible cost (I would expect some things can no longer move out of loops, etc as the control flow becomes much harder to optimize).

Could there be a way that after the first deoptimization, Truffle/Graal learns from it and then continue as if the method was annotated?
This would provide good performance until the first exception is throw, and still not end up in a deopt loop after for all exceptions thrown by the method.

In TruffleRuby, I'd bet hundreds of TruffleBoundary methods would need this flag but don't have it.
It's quite time-consuming to discover such an issue, and we basically only discover them one by one when some code triggers an exception behind a boundary enough times to compile.
To give an idea, there are currently 1037 TruffleBoundary methods in TruffleRuby.
If this cannot be fixed nicely, then maybe we should consider to have throwsControlFlowException/throwsException default to true.

On the good side, it's not so common that exceptions are triggered behind a boundary and such code paths tend to happen not so many times, so it often goes unnoticed. If the exception is a programmer error, then often it is fatal and performance does not matter. Yet, as we execute more code we're very likely too meet more of these cases and that some of them become critical performance bottlenecks.

One also needs to learn that for efficient exception handling, PE needs to see the throw (and the catch), and so only getting the backtrace and building the exception message should be behind a boundary. Also, there needs to be a profile to avoid the exception branch for best performance:
```java
// There can be a Specialization with rewriteOn = ArithmeticException here to avoid explicit checking initially

@Specialization
protected int divide(int a, int b,
                     @Cached("create()") BranchProfile errorProfile) {
    if (b == 0) {
        errorProfile.enter();
        throw new MyLanguageExceptionExtendingControlFlowException(buildMessage(a), getBacktrace());
    }

    // Also need to check that a/b are not a combination of -1/Integer.MIN_VALUE.

    return a / b;
}

@TruffleBoundary
String buildMessage(int a) {
    return String.format("%d cannot be divided by 0", a);
}

@TruffleBoundary
Backtrace getBacktrace() { ... }
```

@arthurp explained this issue on Gitter today:
> I ran into a really surprising (to me) optimization in Truffle and I wanted to check if it is correct.
> It appears that Truffle assumes whenever a `@TruffleBoundary` method throws an exception it should deoptimize instead of calling the exception handler, however Truffle does not remember the exception was thrown so when the code is compiled again it is still missing the exception handler. This behavior is prevented by providing `throwsControlFlowException = true` to `@TruffleBoundary`.
> I have written a small example: https://gist.github.com/arthurp/20d0643d5eb8aadc912640d372e87113

The gist shows detailed examples of throwsControlFlowException including IGV graphs.

**Comment by arthurp** [comment]
@eregon recommended a possible improvement.
> Could there be a way that after the first deoptimization, Truffle/Graal learns from it and then continue as if the method was annotated?

I was wondering how high the cost is expected to be for just having a small branch ending it a deopt. This is what the graph looks like if I use BranchProfile. ([This example is in a new gist.](https://gist.github.com/arthurp/5d5a52ff2e4fb2ac532cf951391f0323)) It seems to me that the cost could be quite low, but I really don't understand the costs of exceptions in Java. Notably if I the throw is inlined (via `allowInlining = true`), the throw and catch is eliminated (unsurprisingly, but it's still good).

![IGV graph with branch profile](https://gist.githubusercontent.com/arthurp/5d5a52ff2e4fb2ac532cf951391f0323/raw/4ba660e56fe1879137bc453f43b91c38ebd4bc1d/RootNode1.5_BeforeLowering.png)

**Comment by chumer** [comment]
> Could there be a way that after the first deoptimization, Truffle/Graal learns from it and then continue as if the method was annotated?

That would be quite difficult but doable, we have set of assumptions per call target where we remember previously failed assumptions about the code. However that might not be desirable in all cases because it would deoptimize and invalidate for an exception that potentially just happened once.

> On the good side, it's not so common that exceptions are triggered behind a boundary and such code paths tend to happen not so many times, so it often goes unnoticed. If the exception is a programmer error, then often it is fatal and performance does not matter. Yet, as we execute more code we're very likely too meet more of these cases and that some of them become critical performance bottlenecks.

I agree with that. I think by default we should deoptimize on exception for boundaries. I think the cases where we expect a ControlFlowException to happen are rare and those can be handled manually. Or are they not rare for Ruby? For normal languages errors its very likely good enough long term to just deoptimize when we see them. Or can you already provide a counter example?

Anyway the documentation is not good enough indeed:

I'd like to suggest a rename to 
```
 public @interface TruffleBoundary {
        boolean transferToInterpreterOnException() default true;
}
```

Which is basically a shortcut for:

```
@TruffleBoundary(transferToInterpreterOnException=false) 
public void boundaryCall() {}

void execute() {
  try {
     boundaryCall();
  } catch (Exception e) {
     CompilerDirectives.transferToInterpreter();
     ...
  }
}
```

If you want to use a profile for your exception you can do:

```
@TruffleBoundary(transferToInterpreterOnException=false) 
public void boundaryCall() {}

final BranchProfile exceptionProfile = BranchProfile.create();
void execute() {
  try {
     boundaryCall();
  } catch (Exception e) {
     exceptionProfile.enter();
     ...
  }
}
```
I don't think a rename is a problem because it is probably rarely used.
What do you think?

**Comment by eregon** [comment]
> That would be quite difficult but doable, we have set of assumptions per call target where we remember previously failed assumptions about the code. However that might not be desirable in all cases because it would deoptimize and invalidate for an exception that potentially just happened once.

It already deoptimizes on every call, so I don't think that's an issue (and invalidates, I just checked, or at least it keeps recompiling the method infinitely if called in a loop).
It would introduce an extra exception path, which in the case the exception no longer happens could have a slight overhead. But I don't think we want to optimize for "exceptions can happen a couple times but never after" at the potential price of deoptimizing all the time if it does not hold.
A BranchProfile doesn't handle this case either, so I don't think this corner case is worth it.

> I think the cases where we expect a ControlFlowException to happen are rare and those can be handled manually.

Not so much, sometimes there is some encoding validation deep behind a boundary and that's supposed to throw a Ruby error (a ControlFlowException), and it's far from trivial to expose the `throw` to PE. Other cases can be just UnsupportedOperationException due to a missing feature or so which gets translated at method boundary to a Ruby error, but still would not matter as maybe it's used as feature detection.
In practice, these kind of errors can happen only once per benchmark iteration, but it's enough to impact performance significantly because of the deopt+invalidate.

> I don't think a rename is a problem because it is probably rarely used.

Agreed.

**Comment by chumer** [comment]
Why is a Ruby error a ControlFlowException? That does not sound right.

> Agreed (although some transition period should exist).

We can do the usual deprecate for 1-2 releases and remove.

**Comment by chumer** [comment]
> It already deoptimizes on every call, so I don't think that's an issue (and invalidates, I just checked, or at least it keeps recompiling the method infinitely if called in a loop).

It definitely should not invalidate. If it does we have another problem.

**Comment by eregon** [comment]
> Why is a Ruby error a ControlFlowException? That does not sound right.

Why not? We want to optimize `begin raise "foo" rescue Exception; end`, and they don't need a Java stacktrace.

**Comment by chumer** [comment]
Why do you think they don't need a java stack trace? A debugger can still observe the exception.

**Comment by smarr** [comment]
Why would a language exception need a Java stack trace?
Doesn't it just need a language stack trace, potentially not even that depending on semantics/guarantees?

**Comment by eregon** [comment]
What @smarr said, we only need the Ruby-level stacktrace.
The Java stacktrace would only be useful for debugging the interpreter, and would incur a significant cost.

**Comment by eregon** [comment]
> It definitely should not invalidate. If it does we have another problem.

I opened an internal issue for that.

**Comment by chumer** [comment]
> Why would a language exception need a Java stack trace? Doesn't it just need a language stack trace, potentially not even that depending on semantics/guarantees?

> The Java stacktrace would only be useful for debugging the interpreter, and would incur a significant cost.

It depends on the usage scenario. In embedded scenarios where host and guest language are used together its helpful to see the java stack trace elements as well interleaved. For normal guest language launchers the java stack trace does not really matter (unless Java interop is used).

Do you have numbers backing this? I don't think the cost is significant compared to capturing language frames. Anyway we can capture the stack frames when a TruffleException reaches the polyglot level or at a Java interop boundary to enhance the stack trace.

**Comment by chumer** [comment]
> We want to optimize begin raise "foo" rescue Exception; end, and they don't need a Java stacktrace.

You need to at least capture the SourceSection where it was thrown, to be able to restore the stack trace for the debugger.

**Comment by arthurp** [comment]
So if exceptions are by default treated as going to the interpreter and slow path, then what is `SlowPathException` for? The documentation states "The Truffle optimizer has special knowledge of this exception class and will never compile a catch block that catches this exception type" which implies to me that other catches can be compiled. I know that throws from boundaries and throws from inside the interpreter are different, but it might be possible to make this all more consistent and easier to understand.

**Comment by arthurp** [comment]
A closely related documentation issue that is probably something we could bundle in here (tell me if you want me to setup a separate issue for this).

`ControlFlowException` explicitly states in it's documentation that Truffle has special knowledge of it. But from what you are all saying, it seems that the only thing that `ControlFlowException` does is prevent the capture of the Java stack (which can also be done with singleton exceptions and other techniques). Is `ControlFlowException` treated specially at all? Or can we get the same performance using any other exception class as long as we avoid stack capture?

**Comment by woess** [comment]
> So if exceptions are by default treated as going to the interpreter and slow path, then what is `SlowPathException` for? The documentation states "The Truffle optimizer has special knowledge of this exception class and will never compile a catch block that catches this exception type" 

As the documentation correctly states, the Truffle optimizer will never compile or even parse a catch block that catches a (subclass of) `SlowPathException` (among a few others), so e.g.:
```java
    try {
        leftNodeValue_ = leftNode_.executeLong(frameValue);
    } catch (UnexpectedResultException ex) {
        Object rightNodeValue = rightNode_.executeGeneric(frameValue);
        return executeAndSpecialize(ex.getResult(), rightNodeValue);
    }
```
is equivalent to:
```java
    try {
        leftNodeValue_ = leftNode_.executeLong(frameValue);
    } catch (UnexpectedResultException ex) {
        CompilerDirectives.transferToInterpreter();
        ...
    }
```

As you suspected, other catch blocks can be compiled, and usually will be unless explicitly cut off using a compiler directive. I say *usually* because Graal can still skip it if the exception was never thrown.

**Comment by woess** [comment]
> Is `ControlFlowException` treated specially at all? Or can we get the same performance using any other exception class as long as we avoid stack capture?

There's some special treatment for it in Truffle when crossing CallTarget boundaries, but as long as you catch it within the same CallTarget you should get the same performance with any other similar exception class or singleton.

**Comment by eregon** [comment]
`throwsControlFlowException` has been deprecated in favor of `transferToInterpreterOnException` https://github.com/graalvm/graal/blob/fe550365bf5d2ef10df9df5c8f0f002615b7cdd9/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/CompilerDirectives.java#L263-L269

I consider the documentation issue fixed, thanks!

But, I still think it would be a good idea to learn automatically from the first exception thrown.
Otherwise, we discover issues one by one after a fair amount of debugging to understand why something deopts. And this repeats on every new application we try to run fast.

**Comment by chumer** [comment]
> But, I still think it would be a good idea to learn automatically from the first exception thrown.
Otherwise, we discover issues one by one after a fair amount of debugging to understand why something deopts. And this repeats on every new application we try to run fast.

Easier said than done. We would add additional overhead for the profile for each boundary call site. Besides we don't have a place to store the profile for boundary calls. Better we give control to the languages about that.

**Comment by eregon** [comment]
What kind of overhead? Should not matter on peak performance. Can we not associate that information with the CallTarget and record it on the first deopt due to an exception from a boundary call? Then the next time Graal compiles it, it looks up if that boundary call should handle exceptions or not.

**Comment by chumer** [comment]
yeah. that could work, but would be a huge performance cliff.

**Comment by eregon** [comment]
In which cases would it be a huge performance cliff?
When it raises a few times but never after?
I think that's a dangerous bet which doesn't hold true in most cases.
If a boundary call fails once, it's very likely to happen again (at that call site) and it's so much better to have an extra exception path than deoptimizing every time.

**Comment by dougxc** [comment]
@chumer , can this issue be closed?

**Comment by chumer** [comment]
Yes. We renamed TruffleBoundary#throwsControlFlowException to TruffleBoundary#transferToInterpeterOnException to make it more clear what it does.
https://github.com/oracle/graal/commit/9b7370be662590c4df291e621ac18cddb3ed9303
